Refactor OSSL_LIB_CTX to avoid using CRYPTO_EX_DATA

hlandau opened the pull request:
This refactors OSSL_LIB_CTX to avoid using CRYPTO_EX_DATA. The assorted objects to be managed by OSSL_LIB_CTX are hardcoded and are initialized eagerly rather than lazily, which avoids the need for locking on access in most cases.

This ended up surprisingly simple. Of course I'm expecting people to throw rocks at this; this can be considered a first draft.

Currently this has the different parts of the code still calling `ossl_lib_ctx_get_data` with vestigial method structures, implemented using a big switch statement. Probably after this we can replace these with separate methods per object to be retrieved and get rid of the indices and most of the method structures (except for the separate fips module). I figured I'd leave this to a separate PR to keep the size of this PR down and make it easier to review.

The 02-test_internal_context test has been removed since it seems to mostly test the dynamic functionality of OSSL_LIB_CTX which no longer exists.

This is built on top of the scrypt KDF dup bugfix #17873 because it's dependent on it.

Fixes #17116.

hlandau commented:
Failures are being investigated.

t8m commented on test/context_internal_test.c:
```diff
@@ -1,137 +0,0 @@
-
-static int test_set0_default(void)
-{
-    OSSL_LIB_CTX *global = OSSL_LIB_CTX_get0_global_default();
```
I suppose we would still want to keep this test to test the OSSL_LIB_CTX_get0_global_default and OSSL_LIB_CTX_set0_default functions.

hlandau replied:
Done.

hlandau commented:
Investigating deadlocks.

slontis reviewed (commented):
Most of the functions that have been made non static may now need to be namespaced (e.g ossl_blah).

hlandau commented:
Updated to add namespacing to symbols.

slontis commented:
Nice.. Code that doesn't use crypto_ex_data seems easier to understand :)
I would be interested to see what performance difference this makes when doing fetching of an algorithm in a loop, after doing an initial prefetch wilth these changes vs before the changes.

paulidale commented:
I'll make an effort to review this properly tomorrow.  Too much to get though today while retaining what little remains of my sanity.

paulidale commented on crypto/context.c:
```diff
@@ -299,127 +514,114 @@ int ossl_lib_ctx_is_global_default(OSSL_LIB_CTX *ctx)
+        return NULL;
+
+    switch (index) {
+        case OSSL_LIB_CTX_PROPERTY_STRING_INDEX:
```
We don't indent the `case` bit and only indent the body once.

qv https://github.com/openssl/technical-policies/blob/master/policies/coding-style.md#chapter-3-placing-braces-and-spaces

paulidale commented on crypto/context_local.h:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2019-2022 The OpenSSL Project Authors. All Rights Reserved.
```
Just 2022

paulidale commented on providers/implementations/kdfs/scrypt.c:
```diff
@@ -56,7 +56,7 @@ typedef struct {
 static void kdf_scrypt_init(KDF_SCRYPT *ctx);
 
-static void *kdf_scrypt_new(void *provctx)
+static void *kdf_scrypt_new_inner(OSSL_LIB_CTX *libctx)
```
Is this caught from a different PR?  It doesn't seem releated.
But it is a fix.

slontis replied:
It should be rebased since #17873 has been merged

paulidale replied:
Thanks.  No merge conflict is showing which is a bit odd.

paulidale reviewed (approved):
A couple of style nits to address but good going.

hlandau commented:
Had to rebase, not sure what happened. Anyway, updated.

mattcaswell commented on crypto/context.c:
```diff
@@ -14,27 +14,45 @@
 
     /*
      * For most data in the OSSL_LIB_CTX we just use ex_data to store it. But
      * that doesn't work for ex_data itself - so we store that directly.
```
This comment seems out of date now

mattcaswell commented on crypto/context.c:
```diff
@@ -80,62 +99,258 @@ static int context_init(OSSL_LIB_CTX *ctx)
+    if (ctx->rand_crngt_lock == NULL)
+        goto err;
 
     /* OSSL_LIB_CTX is built on top of ex_data so we initialise that directly */
```
It isn't really built on top of ex_data any more.

mattcaswell commented on doc/internal/man3/ossl_lib_ctx_get_data.pod:
```diff
@@ -28,90 +28,46 @@ ossl_lib_ctx_is_child
-ossl_lib_ctx_run_once is used to run some initialisation routine I<run_once_fn>
+ossl_lib_ctx_run_once() is used to run some initialisation routine I<run_once_fn>
 exactly once per library context I<ctx> object. Each initialisation routine
 should be allocate a unique run once index in cryptlib.h.
```
Side note (not for this PR): This function is not actually used anywhere in the code. It is a candidate for removal.

mattcaswell commented on crypto/context.c:
```diff
@@ -299,127 +514,114 @@ int ossl_lib_ctx_is_global_default(OSSL_LIB_CTX *ctx)
+        /*
+         * fips is a separate module which may or may not be loaded,
+         * so we have to do this lazily.
+         */
```
I'm not sure why it isn't sufficient to do this eagerly but wrapped in "ifdef FIPS_MODULE".

hlandau replied:
As I recall, trying to initialize the FIPS module eagerly resulted in a rather obtuse issue caused by FIPS trying to call into libctx methods before libctx_init had been completed.

mattcaswell commented on crypto/context.c:
```diff
@@ -299,127 +514,114 @@ int ossl_lib_ctx_is_global_default(OSSL_LIB_CTX *ctx)
+                return NULL;
+
+            if (ctx->fips_prov == NULL) {
+                ctx->fips_prov = meth->new_func(ctx);
```
If this were wrapped in "ifdef FIPS_MODULE" I'm not sure why we need the method? Why can't we just handle it in the same way as all the others? Note: context.c is compiled *twice*. Once with FIPS_MODULE defined and once without. The version without FIPS_MODULE defined is linked into libcrypto.so. The version with FIPS_MODULE defined is linked into fips.so so we should be able to use symbols from that module.

levitte commented on crypto/context_local.h:
```diff
@@ -0,0 +1,44 @@
+/*
```
Please rename this to `include/crypto/context.h` and modify the inclusion lines to `#include "crypto/context.h"`.  Compiler command lines have the necessary flags for this.

levitte commented on crypto/encode_decode/decoder_meth.c:
```diff
@@ -80,8 +81,8 @@ static void *decoder_store_new(OSSL_LIB_CTX *ctx)
 static const OSSL_LIB_CTX_METHOD decoder_store_method = {
     /* We want decoder_store to be cleaned up before the provider store */
     OSSL_LIB_CTX_METHOD_PRIORITY_2,
-    decoder_store_new,
-    decoder_store_free,
+    ossl_decoder_store_new,
+    ossl_decoder_store_free,
 };
```
... and this

levitte commented on crypto/encode_decode/encoder_meth.c:
```diff
@@ -80,8 +81,8 @@ static void *encoder_store_new(OSSL_LIB_CTX *ctx)
 static const OSSL_LIB_CTX_METHOD encoder_store_method = {
     /* We want encoder_store to be cleaned up before the provider store */
     OSSL_LIB_CTX_METHOD_PRIORITY_2,
-    encoder_store_new,
-    encoder_store_free,
+    ossl_encoder_store_new,
+    ossl_encoder_store_free,
 };
```
... and this

levitte commented on crypto/initthread.c:
```diff
@@ -274,15 +275,15 @@ static void *thread_event_ossl_ctx_new(OSSL_LIB_CTX *libctx)
 
 static const OSSL_LIB_CTX_METHOD thread_event_ossl_ctx_method = {
     OSSL_LIB_CTX_METHOD_DEFAULT_PRIORITY,
-    thread_event_ossl_ctx_new,
-    thread_event_ossl_ctx_free,
+    ossl_thread_event_ctx_new,
+    ossl_thread_event_ctx_free,
 };
```
... and this

levitte commented on crypto/property/defn_cache.c:
```diff
@@ -58,14 +59,14 @@ static void property_defns_free(void *vproperty_defns)
 
 static const OSSL_LIB_CTX_METHOD property_defns_method = {
     OSSL_LIB_CTX_METHOD_DEFAULT_PRIORITY,
-    property_defns_new,
-    property_defns_free,
+    ossl_property_defns_new,
+    ossl_property_defns_free,
 };
```
... and this

levitte commented on crypto/property/property_string.c:
```diff
@@ -114,16 +115,16 @@ static void *property_string_data_new(OSSL_LIB_CTX *ctx) {
 
 static const OSSL_LIB_CTX_METHOD property_string_data_method = {
     OSSL_LIB_CTX_METHOD_DEFAULT_PRIORITY,
-    property_string_data_new,
-    property_string_data_free,
+    ossl_property_string_data_new,
+    ossl_property_string_data_free,
 };
```
... and this

levitte commented on crypto/provider_child.c:
```diff
@@ -48,8 +49,8 @@ static void child_prov_ossl_ctx_free(void *vgbl)
 
 static const OSSL_LIB_CTX_METHOD child_prov_ossl_ctx_method = {
     OSSL_LIB_CTX_METHOD_LOW_PRIORITY,
-    child_prov_ossl_ctx_new,
-    child_prov_ossl_ctx_free,
+    ossl_child_prov_ctx_new,
+    ossl_child_prov_ctx_free,
 };
```
... and this

levitte commented on crypto/provider_conf.c:
```diff
@@ -57,8 +58,8 @@ static void prov_conf_ossl_ctx_free(void *vpcgbl)
 static const OSSL_LIB_CTX_METHOD provider_conf_ossl_ctx_method = {
     /* Must be freed before the provider store is freed */
     OSSL_LIB_CTX_METHOD_PRIORITY_2,
-    prov_conf_ossl_ctx_new,
-    prov_conf_ossl_ctx_free,
+    ossl_prov_conf_ctx_new,
+    ossl_prov_conf_ctx_free,
 };
```
... and this

levitte commented on crypto/provider_core.c:
```diff
@@ -322,8 +323,8 @@ static void *provider_store_new(OSSL_LIB_CTX *ctx)
 static const OSSL_LIB_CTX_METHOD provider_store_method = {
     /* Needs to be freed before the child provider data is freed */
     OSSL_LIB_CTX_METHOD_PRIORITY_1,
-    provider_store_new,
-    provider_store_free,
+    ossl_provider_store_new,
+    ossl_provider_store_free,
 };
```
... and this

levitte commented on crypto/rand/rand_lib.c:
```diff
@@ -493,8 +494,8 @@ static void rand_ossl_ctx_free(void *vdgbl)
 
 static const OSSL_LIB_CTX_METHOD rand_drbg_ossl_ctx_method = {
     OSSL_LIB_CTX_METHOD_PRIORITY_2,
-    rand_ossl_ctx_new,
-    rand_ossl_ctx_free,
+    ossl_rand_ctx_new,
+    ossl_rand_ctx_free,
 };
```
... and this

levitte commented on crypto/self_test_core.c:
```diff
@@ -32,23 +33,23 @@ struct ossl_self_test_st
 
 static const OSSL_LIB_CTX_METHOD self_test_set_callback_method = {
     OSSL_LIB_CTX_METHOD_DEFAULT_PRIORITY,
-    self_test_set_callback_new,
-    self_test_set_callback_free,
+    ossl_self_test_set_callback_new,
+    ossl_self_test_set_callback_free,
 };
```
... and this

levitte commented on crypto/store/store_meth.c:
```diff
@@ -83,8 +84,8 @@ static void *loader_store_new(OSSL_LIB_CTX *ctx)
 static const OSSL_LIB_CTX_METHOD loader_store_method = {
     /* We want loader_store to be cleaned up before the provider store */
     OSSL_LIB_CTX_METHOD_PRIORITY_2,
-    loader_store_new,
-    loader_store_free,
+    ossl_loader_store_new,
+    ossl_loader_store_free,
 };
```
... and this

levitte reviewed (changes requested):
If I understand correctly, `OSSL_LIB_CTX_METHOD` isn't necessary any more with this PR.  So you should be able to drop all instances, i.e.:

mattcaswell commented:
> If I understand correctly, OSSL_LIB_CTX_METHOD isn't necessary any more with this PR. So you should be able to drop all instances, i.e.:

It's actually still used for OSSL_LIB_CTX_FIPS_PROV_INDEX (although as per my comments above, I'm not convinced it *needs* to be). Other instances probably can be dropped.

hlandau commented:
Yeah. I left out the cleanup of the OSL_LIB_CTX_METHOD structures initially to keep the diff size down. Updated to clean all this up.

It looks like we can indeed avoid needing to lazily load the FIPS stuff.

Updated.

hlandau commented:
Any idea what's going on here? Not like the file has been changed. https://github.com/openssl/openssl/runs/5719759959?check_suite_focus=true

mattcaswell commented on crypto/evp/evp_fetch.c:
```diff
@@ -23,24 +23,6 @@
-
-static void *evp_method_store_new(OSSL_LIB_CTX *ctx)
-{
-    return ossl_method_store_new(ctx);
```
Sorry, didn't spot this inconsistency before. It is strange that we completely remove `evp_method_store_new` and `evp_method_store_free` in order to call `ossl_method_store_new` directly in context.c. However, we don't do the same for `ossl_encoder_store_new`, `ossl_decoder_store_new` or `ossl_loader_store_new`.

mattcaswell commented:
@paulidale - needs reconfirmation

paulidale commented on crypto/context.c:
```diff
@@ -80,62 +93,261 @@ static int context_init(OSSL_LIB_CTX *ctx)
 
-    if (!ossl_crypto_new_ex_data_ex(ctx, CRYPTO_EX_INDEX_OSSL_LIB_CTX, NULL,
-                                    &ctx->data))
+    /* P2. We want evp_method_store to be cleaned up before the provider store */
```
I'd prefer P2 to be _Highest priority_.
Likewise P1 to be _High priority_.

It took me too long to figure out what was meant.

Below uses _default_ and _low_.  I guess P0 and P-1 would work for them but they are less obvious IMO.

paulidale commented on crypto/context.c:
```diff
@@ -14,27 +14,39 @@
+    void *property_defns;
+    void *global_properties;
+    void *drbg;
+    void *drbg_nonce;
```
Not for this PR but it would be really nice if these fields had a type which the compiler could check....

I know that means not using the one accessor function etc.


We have altogether too many void pointers being cast.  Everything across the provider boundary is a void * and mistakes have been made there.

paulidale reviewed (approved):
Still good.  Some of the comments aren't immediately obvious and a suggestion for future improvements.

paulidale commented:
Merged to master.
